### PR TITLE
Use controller to reconcile k8s secrets 

### LINKED ISSUE #188: Use a controller to reconcile secrets
watch `secretsproviderclass` status `byPod` field

### PR DESCRIPTION
**Pull request by aramase**
**What this PR does / why we need it**:

- Adds a new custom resource `SecretProviderClassPodStatus` to track the status of the mount for the pod
  - The `SecretProviderClassPodStatus` is created with the pod as `ownerReference`

```yaml
apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
kind: SecretProviderClassPodStatus
metadata:
 creationTimestamp: "2020-06-02T00:12:38Z"
 generation: 1
 labels:
   internal.secrets-store.csi.k8s.io/node-name: kind-worker
 name: nginx-secrets-store-inline-0-default-azure
 namespace: default
 ownerReferences:
 - apiVersion: v1
   kind: Pod
   name: nginx-secrets-store-inline-0
   uid: d8771ddf-935a-4199-a20b-f35f71c1d9e7
 resourceVersion: "107107"
 selfLink: /apis/secrets-store.csi.x-k8s.io/v1alpha1/namespaces/default/secretproviderclasspodstatuses/nginx-secrets-store-inline-0-default-azure
 uid: 72a0ecb8-c6e5-41e1-8da1-25e37ec61b26
status:
 mounted: true
 podName: nginx-secrets-store-inline-0
 secretProviderClassName: azure
 targetPath: /var/lib/kubelet/pods/d8771ddf-935a-4199-a20b-f35f71c1d9e7/volumes/kubernetes.io~csi/secrets-store-inline/mount
```
- The secret is created and contains the `SecretProviderClass` as `ownerReference`.
- Removes support for provider parameters in pod spec (ref: https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues/204)

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #188 

**Special notes for your reviewer**:
Design doc: https://docs.google.com/document/d/1EM0Lf-qSdUXdfjOx3AZB9yHOr-lw9_CtoUhR0eTmIwo/edit?usp=sharing

### CODE REVIEW

File: charts/secrets-store-csi-driver/templates/secrets-store.csi.x-k8s.io_secretproviderclasses.yaml [hunk 1]
```diff
@@ -3,17 +3,13 @@
...
 metadata:
-  annotations:
-    controller-gen.kubebuilder.io/version: v0.2.4
   creationTimestamp: null
   name: secretproviderclasses.secrets-store.csi.x-k8s.io
 spec:
   group: secrets-store.csi.x-k8s.io
   names:
     kind: SecretProviderClass
-    listKind: SecretProviderClassList
```
Comment (ritazh) [inline charts/secrets-store-csi-driver/templates/secrets-store.csi.x-k8s.io_secretproviderclasses.yaml:14]: what version of controller-gen are you using? last time I checked v0.2.9 still generates these?

File: config/rbac/secretproviderclasspodstatus_editor_role.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# permissions for end users to edit secretproviderclasspodstatuses.
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: secretproviderclasspodstatus-editor-role
+rules:
+- apiGroups:
+  - secrets-store.csi.x-k8s.io.secrets-store.csi.x-k8s.io
```
Comment (ritazh) [inline config/rbac/secretproviderclasspodstatus_editor_role.yaml:8]: this apiGroup is wrong

File: config/rbac/secretproviderclasspodstatus_editor_role.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  verbs:
+  - create
+  - delete
+  - get
+  - list
+  - patch
+  - update
+  - watch
+- apiGroups:
+  - secrets-store.csi.x-k8s.io.secrets-store.csi.x-k8s.io
```
Comment (ritazh) [inline config/rbac/secretproviderclasspodstatus_editor_role.yaml:20]: same here

File: config/rbac/secretproviderclasspodstatus_editor_role.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# permissions for end users to edit secretproviderclasspodstatuses.
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: secretproviderclasspodstatus-editor-role
```
Comment (ritazh) [inline config/rbac/secretproviderclasspodstatus_editor_role.yaml:5]: do you need this file?

File: controllers/secretproviderclasspodstatus_controller.go [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+			}
+		}
+	}
+
+	if len(errs) > 0 {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	logger.Info("reconcile complete")
+	return ctrl.Result{RequeueAfter: 5 * time.Minute}, nil
```
Comment (ritazh) [inline controllers/secretproviderclasspodstatus_controller.go:213]: why is this queue again after 5mins?

File: deploy/rbac-secretproviderclass.yaml [hunk 6]
```diff
@@ -47,3 +49,24 @@ rules:
...
+  - list
+  - patch
+  - update
+  - watch
+- apiGroups:
+  - secrets-store.csi.x-k8s.io
+  resources:
+  - secretproviderclasspodstatuses/status
+  verbs:
+  - get
```
Comment (ritazh) [inline deploy/rbac-secretproviderclass.yaml:72]: patch and update

File: pkg/secrets-store/nodeserver.go [hunk 7]
```diff
@@ -151,6 +116,33 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 		log.Infof("skipping calling provider as it's mock")
 		return &csi.NodePublishVolumeResponse{}, nil
 	}
+
+	/// TODO: providerName is here for backward compatibility. Will eventually deprecate.
```
Comment (ritazh) [inline pkg/secrets-store/nodeserver.go:120]: we can remove this comment if we want to remove support for `providerName` as part of the`volumeAttributes` of the pod spec. then make sure we add a warning about it in the release note.

File: pkg/secrets-store/utils.go [hunk 8]
```diff
@@ -713,90 +201,50 @@ func getMapFromObjectSpec(object map[string]interface{}, key string) (map[string
...
-			return nil, fmt.Errorf("unknown private key type found while getting key. Only rsa and ecdsa are supported")
-		}
-	}
-	if key, err := x509.ParseECPrivateKey(der); err == nil {
-		derKey, err = x509.MarshalECPrivateKey(key)
-		if err != nil {
-			return nil, err
-		}
+	obj.SetLabels(map[string]string{
+		"internal.secrets-store.csi.k8s.io/node-name": nodeID,
```
Comment (ritazh) [inline pkg/secrets-store/utils.go:237]: should use the same const `internalNodeLabel`

File: pkg/secrets-store/nodeserver.go (hunk 7 above)
Comment (aramase) [reply pkg/secrets-store/nodeserver.go]: Opened an issue to track this - https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues/233

File: controllers/secretproviderclasspodstatus_controller.go (hunk 5 above)
Comment (aramase) [reply controllers/secretproviderclasspodstatus_controller.go:215]: This is done to periodically reconcile and ensure secrets that were created still exists and the owner reference for the spc pod status exists as expected

File: config/rbac/secretproviderclasspodstatus_editor_role.yaml (hunk 4 above)
Comment (aramase) [reply config/rbac/secretproviderclasspodstatus_editor_role.yaml]: Removed it!

File: charts/secrets-store-csi-driver/templates/secrets-store.csi.x-k8s.io_secretproviderclasses.yaml (hunk 1 above)
Comment (aramase) [reply charts/secrets-store-csi-driver/templates/secrets-store.csi.x-k8s.io_secretproviderclasses.yaml]: It was an older version `v0.2.0`. Updated it to `v0.3.0`.

File: Makefile [hunk 9]
```diff
@@ -143,3 +153,30 @@ e2e-azure: install-driver
 .PHONY: e2e-vault
 e2e-vault: install-driver
 	bats -t test/bats/vault.bats
+
+# Update CRDs in Helm Chart
+release: manifests
+	cp config/crd/bases/* charts/secrets-store-csi-driver/templates
+	cp config/rbac/* charts/secrets-store-csi-driver/templates
```
Comment (ritazh) [inline Makefile:160]: we also need to update the manifest in `deploy/`

File: controllers/secretproviderclasspodstatus_controller.go (hunk 5 above)
Comment (ritazh) [reply controllers/secretproviderclasspodstatus_controller.go:218]: oh duh. this is the loop :)

File: test/bats/azure.bats [hunk 10]
```diff
@@ -55,6 +55,7 @@ setup() {
 }
 
 @test "CSI inline volume test" {
+  skip
```
Comment (ritazh) [inline test/bats/azure.bats:58]: If we want to skip these tests, should we just remove them?

Comment (aramase) [reply test/bats/azure.bats]: Makes sense! Removed it.

**Pull request merged**